File: Java/opentraceability/src/main/java/opentraceability/models/masterdata/kdes/MasterDataKDEString.java
package opentraceability.models.masterdata.kdes;

import opentraceability.interfaces.IMasterDataKDE;
import opentraceability.utility.XElement;
import org.json.JSONObject;
import org.w3c.dom.Document;
import org.w3c.dom.Element;

import javax.xml.parsers.DocumentBuilderFactory;

import java.util.HashMap;
import java.util.Map;

public class MasterDataKDEString extends IMasterDataKDE {

    public Class valueType = String.class;

    public String value = null;
    public String type = null;
    public Map<String, String> attributes = new HashMap<>();

    public MasterDataKDEString() {
    }

    public MasterDataKDEString(Class valueType) {
        // Default constructor
        this.valueType = valueType;
    }

    public MasterDataKDEString(String ns, String name) {
        super.namespace = ns;
        super.name = name;
    }

    public JSONObject getJson() {
        if (value == null || value.isEmpty()) {
            return null;
        } else {
            return new JSONObject(value);
        }
    }


    public Element getXml() throws Exception {
        if (value == null || value.isEmpty()) {
            return null;
        } else {
            String xname = namespace + name;
            Document document = DocumentBuilderFactory.newInstance().newDocumentBuilder().newDocument();
            Element element = document.createElement(xname);
            element.setTextContent(value);

            // Set the xsi type...
            for (Map.Entry<String, String> entry : attributes.entrySet()) {
                String key = entry.getKey();
                String val = entry.getValue();
                element.setAttribute(key, val);
            }

            return element;
        }
    }


    public void setFromJson(JSONObject json) {
        value = json.toString();
    }

    public void setFromXml(Element xml) {
        value = xml.getTextContent();

        for (int i = 0; i < xml.getAttributes().getLength(); i++) {
            String key = xml.getAttributes().item(i).getNodeName();
            String val = xml.getAttributes().item(i).getNodeValue();
            attributes.put(key, val);
        }
    }

    @Override
    public String toString() {
        return value == null ? "" : value;
    }

    @Override
    public void setFromGS1WebVocabJson(Object json) {
        value = json.toString();
    }

    public Object getGS1WebVocabJson() {
        if (value != null) {
            return value;
        } else {
            return null;
        }
    }

    @Override
    public void setFromEPCISXml(XElement xml) throws Exception {
        name = xml.Attribute("id");
        value = xml.getValue();
    }

    @Override
    public XElement getEPCISXml() throws Exception {
        if (value != null) {
            XElement element = new XElement("attribute");
            element.SetAttributeValue("id", name);
            element.setValue(value);
            return element;
        } else {
            return null;
        }
    }
}

File: Java/OpenTraceability/src/main/opentraceability/models/masterdata/kdes/masterdatakdestring.kt
package models.masterdata.kdes

import interfaces.IMasterDataKDE


class MasterDataKDEString : MasterDataKDEBase, IMasterDataKDE {
    override var ValueType: Class<*>
        get() = String::class.java

    var value: String? = null
    var type: String? = null
    val attributes: MutableMap<String, String> = mutableMapOf()

    constructor(ValueType: Class<*>) {
        // Default constructor
        this.ValueType = ValueType
    }

    constructor(ns: String, name: String) {
        super.namespace = ns
        super.name = name
    }

    fun getJson(): JToken? {
        return if (value.isNullOrBlank()) {
            null
        } else {
            JToken.fromObject(value)
        }
    }

    fun getXml(): XElement? {
        return if (value.isNullOrBlank()) {
            null
        } else {
            val xname = (XNamespace)namespace + super.name
            val x = XElement(xname, value)

            // set the xsi type...
            attributes.forEach { (key, value) ->
                x.add(XAttribute(key, value))
            }

            x
        }
    }

    fun setFromJson(json: JToken) {
        value = json.toString()
    }

    fun setFromXml(xml: XElement) {
        value = xml.value

        xml.attributes().forEach { xatt ->
            attributes[xatt.name.toString()] = xatt.value
        }
    }

    override fun toString(): String {
        return value ?: ""
    }

    override fun setFromGS1WebVocabJson(json: JToken) {
        value = json.toString()
    }

    override fun getGS1WebVocabJson(): JToken? {
        return if (value != null) {
            JToken.fromObject(value)
        } else {
            null
        }
    }

    override fun setFromEPCISXml(xml: XElement) {
        name = xml.attribute("id")?.value ?: ""
        value = xml.value
    }

    override fun getEPCISXml(): XElement? {
        return if (value != null) {
            val x = XElement("attribute")
            x.add(XAttribute("id", super.name))
            x.value = value
            x
        } else {
            null
        }
    }
}
